packages/tools/src/tools/segmentation/PaintFillTool.ts

```diff
@@ -77,18 +85,64 @@ class PaintFillTool extends BaseTool {
     }
 
     const { segmentationId, type } = activeSegmentationRepresentation;
+    if (type === SegmentationRepresentations.Contour) {
+      throw new Error('Not implemented yet');
+    }
+
     const segmentIndex =
       segmentIndexController.getActiveSegmentIndex(segmentationId);
     const segmentsLocked: number[] =
       segmentLocking.getLockedSegments(segmentationId);
     const { representationData } = getSegmentation(segmentationId);
 
-    const { volumeId } = representationData[type] as LabelmapSegmentationData;
-    const segmentation = cache.getVolume(volumeId);
-    const { dimensions, direction } = segmentation;
-    const scalarData = segmentation.getScalarData();
+    const labelmapData =
+      representationData[SegmentationRepresentations.Labelmap];
+
+    let dimensions, direction, scalarData, imageData;
+    const segmentationType = labelmapData.type;
+    const auxiliaryData = { type: '', imageIds: [] };
+    if (segmentationType === 'volume' || !segmentationType) {
+      const { volumeId } = representationData[
+        type
+      ] as LabelmapSegmentationDataVolume;
+      const segmentation = cache.getVolume(volumeId);
+      dimensions = segmentation.dimensions;
+      direction = segmentation.direction;
+      scalarData = segmentation.getScalarData();
+      imageData = segmentation.imageData;
+    } else {
+      const { referencedImageIds } =
+        labelmapData as LabelmapSegmentationDataStack;
+      const {
+        zSpacing,
+        sortedImageIds,
+        origin,
+        direction: calculatedDirection,
+      } = sortImageIds(referencedImageIds);
+      const currentImageId = viewport.getCurrentImageId();
+
+      imageData = createVTKImageDataFromImageId(currentImageId);
+      // hack the imageData so pointInSurroundingSphereCallback can treat it like a volume
+
+      // Hacking the dimensions
+      dimensions = imageData.getDimensions();
+      dimensions[2] = sortedImageIds.length;
+      imageData.setDimensions(dimensions);
+
+      // Hacking the spacing information.
+      const spacing = imageData.getSpacing();
+      spacing[2] = zSpacing;
+      imageData.setSpacing(spacing);
+      // Hacking the origin
+      imageData.setOrigin(origin);
+      direction = calculatedDirection;
+      auxiliaryData.imageIds = sortedImageIds;
+      const image = cache.getDerivedImage(currentImageId);
```

Comment:
you don't need this cache.getDerivedImage. just put a enum/constant in the tools as segmentation prefix and just do `${segPrefix}:${imageId}`